# Project Notebook: The Linear Regression Model

## 1. Introduction

We started by building intuition for model based learning, explored how the linear regression model worked, understood how the two different approaches to model fitting worked, and some techniques for cleaning, transforming, and selecting features. In this project, you can practice what you learned by exploring ways to improve the models we built.

You'll work with housing data for the city of Ames, Iowa, United States from 2006 to 2010. You can also read about the different columns in the data [here](https://s3.amazonaws.com/dq-content/307/data_description.txt).

Let's start by setting up a pipeline of functions that will let us quickly iterate on different models.

**Tasks**

1. Import pandas, matplotlib, and numpy into the environment. Import the classes you need from scikit-learn as well.
2. Read `AmesHousing.tsv` () into a pandas data frame.
3. For the following functions, we recommend creating them in the first few cells in the notebook. This way, you can add cells to the end of the notebook to do experiments and update the functions in these cells.
* Create a function named `transform_features()` that, for now, just returns the train data frame.
* Create a function named `select_features()` that, for now, just returns the Gr Liv Area and SalePrice columns from the train data frame.
* Create a function named `train_and_test()` that, for now:

1. Selects the first 1460 rows from from data and assign to train.
2. Selects the remaining rows from data and assign to test.
3. Trains a model using all numerical columns except the SalePrice column (the target column) from the data frame returned from `select_features()`
4. Tests the model on the test set and returns the `RMSE` value.

In [175]:
# import pandas and prerequisite classes
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import seaborn as sns


In [176]:
#Read AmesHousing.tsv () into a pandas data frame.
data = pd.read_csv('https://bit.ly/3boZCX4', delimiter="\t")

In [177]:
data.head(5)

Order        PID  MS SubClass  ... Sale Type  Sale Condition  SalePrice
0      1  526301100           20  ...       WD           Normal     215000
1      2  526350040           20  ...       WD           Normal     105000
2      3  526351010           20  ...       WD           Normal     172000
3      4  526353030           20  ...       WD           Normal     244000
4      5  527105010           60  ...       WD           Normal     189900

[5 rows x 82 columns]



**Uncomment to run**

In [178]:
#Create a function named transform_features() that, for now, just returns the train data frame.
def transform_features(data):
    return data

transform_features

<function __main__.transform_features>

In [179]:
#Create a function named select_features() that, for now, just returns the Gr Liv Area and SalePrice columns from the train data frame.
def select_features(data):
    return data[['Gr Liv Area', 'SalePrice']]

select_features



<function __main__.select_features>

In [180]:
# Create a function named train_and_test() that, for now:
#Trains a model using all numerical columns except the SalePrice column (the target column) from the data frame returned from select_features()
# Tests the model on the test set and returns the RMSE value.

def train_and_test(data):
#Using select_dtypes to select only numerical columns 
    train= data[0:1460]
    test =data[1460:]

    train_numerical_columns = train.select_dtypes(include = ['float','integer'])
    test_numerical_columns = test.select_dtypes(include = ['float', 'integer'])

    # #dropping the null values in the train and test numerical cols
    # train_numerical_columns = train_numerical_columns.dropna()
    # test_numerical_columns = test_numerical_columns.dropna()

  # Using pd.drop to drop the target column 'SalePrice' from the train and test data
    train_features = train_numerical_columns.drop(['SalePrice'], axis=1)
    test_features = test_numerical_columns.drop(['SalePrice'], axis=1)

    
# # Identifying the target column 'SalePrice' from the train and test data   
    train_target = train_numerical_columns['SalePrice']
    test_target = test_numerical_columns['SalePrice']

# #Creating a linear regregersiion instanace

    linear_model = LinearRegression()

# # Training the model
    linear_model.fit(train_features, train_target)


# # Predicting using the test set
    predictions = linear_model.predict(test_features)

# # Calculating the RMSE   
    mse = mean_squared_error(test_target, predictions)
    rmse = np.sqrt(mse)
    return rmse




In [181]:
transform_data = transform_features(data)
filtered_data = select_features(transform_data)
rmse = train_and_test(filtered_data)
rmse

57088.25161263909

## 2. Feature Engineering

Let's now start removing features with many missing values, diving deeper into potential categorical features, and transforming text and numerical columns. Update `transform_features()` so that any column from the data frame with more than 25% (or another cutoff value) missing values is dropped. You also need to remove any columns that leak information about the sale (e.g. like the year the sale happened). In general, the goal of this function is to:

* remove features that we don't want to use in the model, just based on the number of missing values or data leakage.
* transform features into the proper format (numerical to categorical, scaling numerical, filling in missing values, etc).
* create new features by combining other features.

Next, you need to get more familiar with the remaining columns by reading the data documentation for each column, determining what transformations are necessary (if any), and more. As we mentioned earlier, succeeding in predictive modeling (and competitions like Kaggle) is highly dependent on the quality of features the model has. Libraries like scikit-learn have made it quick and easy to simply try and tweak many different models, but cleaning, selecting, and transforming features are still more of an art that requires a bit of human ingenuity.

**Tasks**

1. As we mentioned earlier, we recommend adding some cells to explore and experiment with different features (before rewriting these functions).

2. The `transform_features()` function shouldn't modify the train data frame and instead return a new one entirely. This way, we can keep using train in the experimentation cells.

3. Which columns contain less than 5% missing values?
* For numerical columns that meet this criteria, let's fill in the missing values using the most popular value for that column.

4. What new features can we create, that better capture the information in some of the features?
* An example of this would be the `years_until_remod` feature we created in the last lesson.

5. Which columns need to be dropped for other reasons?
* Which columns aren't useful for machine learning?
* Which columns leak data about the final sale?

#### Updating the transform function with feature engineering to see how the model improves

**Uncomment to run**

In [ ]:
# #Read AmesHousing.tsv () into a pandas data frame.
# data = pd.read_csv('https://bit.ly/3boZCX4', delimiter="\t")

# def transform_features(data):

#     # Which columns contain less than 5% missing values? ie ratio of missing values < 5%
#     # percentage of missing values in each column ie # ratio of missing values = (Number of missing qty/ number of observations )*100
#     # saving missing values columns in a variable named percentage_missing
#     percentage_missing = data.isnull().sum()/len(data)*100

#     # saving column names in a variable named all_cols
#     all_cols = data.columns

#     # Which columns contain less than 5% missing values?
#     # Instanciating a new variable to store column names having missing values less than the threshold ie have missing values <5%
#     below_5 = [ ]

#     for i in range(data.columns.shape[0]):
#         if percentage_missing[i]<=5: #setting the missing threshold at less than or equal to 5%
#             below_5.append(all_cols[i])

#     # creating a new dataframe using the above variables
#     new_data = data[below_5]
#     new_data.head()

#     #For numerical columns that contain less than 5% missing values, fill their missing values using the most popular value for that column.
#     ## Get the numerical columns from new_data df
#     numerical_cols = new_data.columns[new_data.dtypes!='object']

#     #function to fill missing values with the mode for the numerical columns in new_data df
#     fill_mode = lambda col: col.fillna(col.mode())

#     #apply the fill_mode function to fill the missing values
#     new_data[numerical_cols]=new_data[numerical_cols].apply(fill_mode)


#     #Creating new features that better capture the information in some of the features # An example of this would be the years_until_remod feature we created in the last lesson.
#     #Feature: How old was the property when it was sold? insert a new column in new_data df with the value of the property age
#     new_data['years_until_sold'] = new_data['Yr Sold'] - new_data['Year Built']

#     #dropping rows with negative qty on 'years_until_sold' column
#     new_data.drop(new_data[new_data['years_until_sold'] < 0].index, inplace = True)


#     # Feature: How long after selling was the property remodelled?
#     new_data['years_remod_after_sale'] = new_data['Yr Sold'] - new_data['Year Remod/Add']

#     #dropping rows with a negative qty on 'years_remod_after_sale' column
#     new_data.drop(new_data[new_data['years_remod_after_sale'] < 0].index, inplace = True)


#     # Feature : Age of the property when it was remodelled?
#     new_data['years_until_remoddeling'] = new_data['Year Remod/Add'] - new_data['Year Built']

#     #dropping the rows with negative qty on 'years_until_remoddeling' column
#     new_data.drop(new_data[new_data['years_until_remoddeling'] < 0].index, inplace = True)

#     # Dropping columns which aren't useful for machine learning,and leak data about the final sale
#     columns_to_drop = ['Order','PID','Year Built','Year Remod/Add','Yr Sold', 'Mo Sold', 'Sale Type'
#                         , 'Sale Condition' ]

#     new_data.drop(columns_to_drop, axis = 1, inplace = True)

#     return new_data


#  #Calling the select_features function that returns the Gr Liv Area and SalePrice columns from the train data frame.
# def select_features(data):
#     return data[['Gr Liv Area', 'SalePrice']]

# select_features 


# # Calling the train_and_test function that trains the model using all numerical columns except the SalePrice column (the target column)
# # from the data frame returned from select_features(), Tests the model on the test set and returns the RMSE value.

# def train_and_test(data):
# #Using select_dtypes to select only numerical columns 
#     train= data[0:1460]
#     test =data[1460:]

#     train_numerical_columns = train.select_dtypes(include = ['float','integer'])
#     test_numerical_columns = test.select_dtypes(include = ['float', 'integer'])

#     # #dropping the null values in the train and test numerical cols
#     # train_numerical_columns = train_numerical_columns.dropna()
#     # test_numerical_columns = test_numerical_columns.dropna()

#   # Using pd.drop to drop the target column 'SalePrice' from the train and test data
#     train_features = train_numerical_columns.drop(['SalePrice'], axis=1)
#     test_features = test_numerical_columns.drop(['SalePrice'], axis=1)

    
# # # Identifying the target column 'SalePrice' from the train and test data   
#     train_target = train_numerical_columns['SalePrice']
#     test_target = test_numerical_columns['SalePrice']

# # # #Creating a linear regression instance  

#     linear_model = LinearRegression()

# # # Training the model
#     linear_model.fit(train_features, train_target)


# # # Predicting using the test set
#     predictions = linear_model.predict(test_features)

# # # Calculating the RMSE   
#     mse = mean_squared_error(test_target, predictions)
#     rmse = np.sqrt(mse)
#     return rmse

# transform_data = transform_features(data)
# filtered_data = select_features(transform_data)
# rmse = train_and_test(filtered_data)
# rmse



The RMSE score improved from 57088.25161263909 to 55284.62277814025


## 3. Feature Selection

Now that we have cleaned and transformed a lot of the features in the data set, it's time to move on to feature selection for numerical features.

**Tasks**

1. Generate a correlation heatmap matrix of the numerical features in the training data set.
* Which features correlate strongly with our target column, `SalePrice`?
* Calculate the correlation coefficients for the columns that seem to correlate well with `SalePrice`. Because we have a pipeline in place, it's easy to try different features and see which features result in a better cross validation score.

2. Which columns in the data frame should be converted to the categorical data type? All of the columns that can be categorized as nominal variables are candidates for being converted to categorical. Here are some other things you should think about:
* If a categorical column has hundreds of unique values (or categories), should you keep it? When you dummy code this column, hundreds of columns will need to be added back to the data frame.
* Which categorical columns have a few unique values but more than 95% of the values in the column belong to a specific category? This would be similar to a low variance numerical feature (no variability in the data for the model to capture).

3. Which columns are currently numerical but need to be encoded as categorical instead (because the numbers don't have any semantic meaning)?

4. What are some ways we can explore which categorical columns "correlate" well with `SalePrice`?

5. Update the logic for the `select_features()` function. This function should take in the new, modified train and test data frames that were returned from `transform_features()`.

In [ ]:
# # # # Your code goes here


# #Read AmesHousing.tsv () into a pandas data frame.
# data = pd.read_csv('https://bit.ly/3boZCX4', delimiter="\t")


# # # Generate a correlation heatmap matrix of the numerical features in the training data set. 
# train= data[0:1460]
# test =data[1460:]
# train_numerical_df = train.select_dtypes(include = ['float','integer'])
# train_numerical_df.corr()['SalePrice'].abs().sort_values(ascending = False)

In [ ]:
# # Calculate the correlation coefficients for the columns that seem to correlate well with SalePrice
# #create new variable for correlation coefficient values greater than 0.7 since a relationship between two variables is generally considered strong when their r value is larger than 0.7
# coef_saleprice = train_numerical_df.corr()['SalePrice'].abs().sort_values(ascending = False)
# high_coef = coef_saleprice[coef_saleprice > 0.5].index
# sns.heatmap(train_numerical_df[high_coef].corr().abs())


It looks like following columns are strongly related to each other.

Total Bsmt SF and 1st Flr SF
TotRms AbvGrd and Gr Liv Area
Garage Cars and Garage Area




In [ ]:
# # Below are the columns with categorical values /categorical data type? 
# # All of the columns that can be categorized as nominal variables are candidates for being converted to categorical. 

# train_categorical_cols = train.columns[train.dtypes=='object'].
# train_categorical_cols



---


It looks like following columns are categorical in nature.

'MS Zoning', 'Street', 'Alley', 'Lot Shape', 'Land Contour',
       'Utilities', 'Lot Config', 'Land Slope', 'Neighborhood', 'Condition 1',
       'Condition 2', 'Bldg Type', 'House Style', 'Roof Style', 'Roof Matl',
       'Exterior 1st', 'Exterior 2nd', 'Mas Vnr Type', 'Exter Qual',
       'Exter Cond', 'Foundation', 'Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure',
       'BsmtFin Type 1', 'BsmtFin Type 2', 'Heating', 'Heating QC',
       'Central Air', 'Electrical', 'Kitchen Qual', 'Functional',
       'Fireplace Qu', 'Garage Type', 'Garage Finish', 'Garage Qual',
       'Garage Cond', 'Paved Drive', 'Pool QC', 'Fence', 'Misc Feature',
       'Sale Type', 'Sale Condition


We shall drop columns that  have more than 10 unique values before dummy coding the rest of the columns

In [ ]:
# #checking the added columns after encoding
# print(transform_data.shape)



---

the dataframe transform_data now has 492 columns after adding the dummy columns

In [ ]:
# # Update the logic for the select_features() function. 
# This function should take in the new, modified train and test data frames that were returned from transform_features().

#### Updating the select function with feature selection to see how the model improves

**Uncomment to run**

In [ ]:
 #Calling the select_features function and updating it with feature engineering done.

def select_features(data):
    train_numerical_df = train.select_dtypes(include = ['float','integer'])
    train_numerical_df.corr()['SalePrice'].abs().sort_values(ascending = False)
    coef_saleprice = train_numerical_df.corr()['SalePrice'].abs().sort_values(ascending = False)
    high_coef = coef_saleprice[coef_saleprice > 0.5].index

    train_categorical_cols = train.columns[train.dtypes=='object']
    train_categorical_cols

# testing which categorical columns have <10 unique values.
    transform_categorical_cols = []
    for col in train_categorical_cols:
        if col in transform_data.columns:
            transform_categorical_cols.append(col)

# Geting the unique values in each categorical column
    uniqueness_counts = transform_data[transform_categorical_cols].apply(lambda col: len(col.value_counts())).sort_values()

# Seleting all the categorical columns with a count >10 unique values and assignning them a variable to be used in dropping
    drop_nonuniq_cols = uniqueness_counts[uniqueness_counts > 10].index

#updating the transform_data df after dropping these columns
    transform_data = transform_data.drop(drop_nonuniq_cols, axis=1)
 
    text_cols = transform_data.select_dtypes(include=['object'])
    for col in text_cols:
      transform_data[col] = transform_data[col].astype('category')
    
# Create dummy columns and adding them back to the transform_data df
    transform_data = pd.concat([
        transform_data, 
        pd.get_dummies(transform_data.select_dtypes(include=['category']))
    ], axis=1)

    
    return data

## 4. Train and Test

Now for the final part of the pipeline, training and testing. When iterating on different features, using simple validation is a good idea. Let's add a parameter named `k` that controls the type of cross validation that occurs.

**Tasks**

1. The optional `k` parameter should accept integer values, with a default value of `0`.

2. When `k` equals `0`, perform holdout validation (what we already implemented):

* Select the first `1460` rows and assign to `train`.
* Select the remaining rows and assign to test.
* Train on `train` and `test` on test.
* Compute the `RMSE` and return.

3. When k equals 1, perform simple cross validation:

* Shuffle the ordering of the rows in the data frame.
* Select the first 1460 rows and assign to `fold_one`.
* Select the remaining rows and assign to `fold_two`.
* Train on `fold_one` and test on `fold_two`.
* Train on `fold_two` and test on `fold_one`.
* Compute the average RMSE and return.

4. When `k` is greater than `0`, implement k-fold cross validation using `k` folds:

* Perform `k-fold` cross validation using k folds.
* Calculate the average `RMSE` value and return this value.

#### Updating the train and test function with cross validation

**Uncomment to run**

In [ ]:
# def train_and_test(data, k=0):

# # initialize count to 0
#   if k == 0:
#     train= data[0:1460]
#     test =data[1460:]
# #Using select_dtypes to select only numerical columns 

#     train_numerical_columns = train.select_dtypes(include = ['float','integer'])
#     test_numerical_columns = test.select_dtypes(include = ['float', 'integer'])

#   # Using pd.drop to drop the target column 'SalePrice' from the train and test data
#     train_features = train_numerical_columns.drop(['SalePrice'], axis=1)
#     test_features = test_numerical_columns.drop(['SalePrice'], axis=1)

# # # Identifying the target column 'SalePrice' from the train and test data   
#     train_target = train_numerical_columns['SalePrice']
#     test_target = test_numerical_columns['SalePrice']

# #  #Creating a linear regression instance  

#     linear_model = LinearRegression()

# # # Training the model
#     linear_model.fit(train_features, train_target)


# # # Predicting using the test set
#     predictions = linear_model.predict(test_features)

# # # Calculating the RMSE   
#     mse = mean_squared_error(test_target, predictions)
#     rmse = np.sqrt(mse)
#     return rmse 


# # When k equals 1, perform simple cross validation:
#   if k == 1:
#         # shuffling all rows in the df
#         shuffled_df = data.sample(frac=1, )
#         train = data[:1460]
#         test = data[1460:]
        

#         train_numerical_columns = train.select_dtypes(include = ['float','integer'])
#         test_numerical_columns = test.select_dtypes(include = ['float', 'integer'])

#       # Using pd.drop to drop the target column 'SalePrice' from the train and test data
#         train_features = train_numerical_columns.drop(['SalePrice'], axis=1)
#         test_features = test_numerical_columns.drop(['SalePrice'], axis=1)

#     # # Identifying the target column 'SalePrice' from the train and test data   
#         train_target = train_numerical_columns['SalePrice']
#         test_target = test_numerical_columns['SalePrice']

#       # #Creating a linear regression instance  
#         linear_model = LinearRegression()  
#         linear_model.fit(train_features, train_target)
#         predictions_one = linear_model.predict(test_features)        
        
#         mse_one = mean_squared_error(test_target, predictions_one)
#         rmse_one = np.sqrt(mse_one)
        
#         linear_model.fit(train_features, train_target)
#         predictions_two = linear_model.predict(test_features)        
       
#         mse_two = mean_squared_error(test_target, predictions_two)
#         rmse_two = np.sqrt(mse_two)
        
#         avg_rmse = np.mean([rmse_one, rmse_two])
#         print(rmse_one)
#         print(rmse_two)
#         return avg_rmse

#   else:
#         kf = KFold(n_splits=k, shuffle=True)
#         rmse_values = []
#         for train_index, test_index, in kf.split(data):
#             train = data.iloc[train_index]
#             test = data.iloc[test_index]
#             linear_model.fit(train_features, train_target)
#             predictions = linear_model.predict(test_features)
         

#         mse = mean_squared_error(test_target, predictions)
#         rmse = np.sqrt(mse)
#         rmse_values.append(rmse)
#         print(rmse_values)
#         avg_rmse = np.mean(rmse_values)
#         return avg_rmse
          


## 5. Model Testing

#### Combining all the  functions with to see how our model has improved.

In [200]:
#Read AmesHousing.tsv () into a pandas data frame.
data = pd.read_csv('https://bit.ly/3boZCX4', delimiter="\t")

#####Transform_features function

In [201]:
def transform_features(data):

    # Which columns contain less than 5% missing values? ie ratio of missing values < 5%
    # percentage of missing values in each column ie # ratio of missing values = (Number of missing qty/ number of observations )*100
    # saving missing values columns in a variable named percentage_missing
    percentage_missing = data.isnull().sum()/len(data)*100

    # saving column names in a variable named all_cols
    all_cols = data.columns

    # Which columns contain less than 5% missing values?
    # Instanciating a new variable to store column names having missing values less than the threshold ie have missing values <5%
    below_5 = [ ]

    for i in range(data.columns.shape[0]):
        if percentage_missing[i]<=5: #setting the missing threshold at less than or equal to 5%
            below_5.append(all_cols[i])

    # creating a new dataframe using the above variables
    new_data = data[below_5]
    new_data.head()

    #For numerical columns that contain less than 5% missing values, fill their missing values using the most popular value for that column.
    ## Get the numerical columns from new_data df
    numerical_cols = new_data.columns[new_data.dtypes!='object']

    #function to fill missing values with the mode for the numerical columns in new_data df
    fill_mode = lambda col: col.fillna(col.mode())

    #apply the fill_mode function to fill the missing values
    new_data[numerical_cols]=new_data[numerical_cols].apply(fill_mode)


    #Creating new features that better capture the information in some of the features # An example of this would be the years_until_remod feature we created in the last lesson.
    #Feature: How old was the property when it was sold? insert a new column in new_data df with the value of the property age
    new_data['years_until_sold'] = new_data['Yr Sold'] - new_data['Year Built']

    #dropping rows with negative qty on 'years_until_sold' column
    new_data.drop(new_data[new_data['years_until_sold'] < 0].index, inplace = True)


    # Feature: How long after selling was the property remodelled?
    new_data['years_remod_after_sale'] = new_data['Yr Sold'] - new_data['Year Remod/Add']

    #dropping rows with a negative qty on 'years_remod_after_sale' column
    new_data.drop(new_data[new_data['years_remod_after_sale'] < 0].index, inplace = True)


    # Feature : Age of the property when it was remodelled?
    new_data['years_until_remoddeling'] = new_data['Year Remod/Add'] - new_data['Year Built']

    #dropping the rows with negative qty on 'years_until_remoddeling' column
    new_data.drop(new_data[new_data['years_until_remoddeling'] < 0].index, inplace = True)

    # Dropping columns which aren't useful for machine learning,and leak data about the final sale
    columns_to_drop = ['Order','PID','Year Built','Year Remod/Add','Yr Sold', 'Mo Sold', 'Sale Type'
                        , 'Sale Condition' ]

    new_data.drop(columns_to_drop, axis = 1, inplace = True)

    return new_data  

 

#####Select_features function

In [202]:
#Calling the select_features function and updating it with feature engineering done.

def select_features(new_data):
    train_numerical_df = train.select_dtypes(include = ['float','integer'])
    train_numerical_df.corr()['SalePrice'].abs().sort_values(ascending = False)
    coef_saleprice = train_numerical_df.corr()['SalePrice'].abs().sort_values(ascending = False)
    high_coef = coef_saleprice[coef_saleprice > 0.5].index

    train_categorical_cols = train.columns[train.dtypes=='object']
    train_categorical_cols

# testing which categorical columns have <10 unique values.
    transform_categorical_cols = []
    for col in train_categorical_cols:
        # if col in transform_data.columns:
            transform_categorical_cols.append(col)

# Geting the unique values in each categorical column
    uniqueness_counts = new_data[transform_categorical_cols].apply(lambda col: len(col.value_counts())).sort_values()

# Seleting all the categorical columns with a count >10 unique values and assignning them a variable to be used in dropping
    drop_nonuniq_cols = uniqueness_counts[uniqueness_counts > 10].index

#updating the transform_data df after dropping these columns
    transform_data = new_data.drop(drop_nonuniq_cols, axis=1)
 
    text_cols = new_data.select_dtypes(include=['object'])
    for col in text_cols:
      new_data[col] = new_data[col].astype('category')
    
# Create dummy columns and adding them back to the transform_data df
    new_data = pd.concat([
        new_data, 
        pd.get_dummies(new_data.select_dtypes(include=['category']))
    ], axis=1)

    
    return data

#####Train_test function

In [203]:
def train_and_test(data, k=0):

# initialize count to 0
  if k == 0:
    train= data[0:1460]
    test =data[1460:]
#Using select_dtypes to select only numerical columns 

    train_numerical_columns = train.select_dtypes(include = ['float','integer'])
    test_numerical_columns = test.select_dtypes(include = ['float', 'integer'])

  # Using pd.drop to drop the target column 'SalePrice' from the train and test data
    train_features = train_numerical_columns.drop(['SalePrice'], axis=1)
    test_features = test_numerical_columns.drop(['SalePrice'], axis=1)

# # Identifying the target column 'SalePrice' from the train and test data   
    train_target = train_numerical_columns['SalePrice']
    test_target = test_numerical_columns['SalePrice']

#  #Creating a linear regression instance  

    linear_model = LinearRegression()

# # Training the model
    linear_model.fit(train_features, train_target)


# # Predicting using the test set
    predictions = linear_model.predict(test_features)

# # Calculating the RMSE   
    mse = mean_squared_error(test_target, predictions)
    rmse = np.sqrt(mse)
    return rmse 


# When k equals 1, perform simple cross validation:
  if k == 1:
        # shuffling all rows in the df
        shuffled_df = data.sample(frac=1, )
        train = data[:1460]
        test = data[1460:]
        

        train_numerical_columns = train.select_dtypes(include = ['float','integer'])
        test_numerical_columns = test.select_dtypes(include = ['float', 'integer'])

      # Using pd.drop to drop the target column 'SalePrice' from the train and test data
        train_features = train_numerical_columns.drop(['SalePrice'], axis=1)
        test_features = test_numerical_columns.drop(['SalePrice'], axis=1)

    # # Identifying the target column 'SalePrice' from the train and test data   
        train_target = train_numerical_columns['SalePrice']
        test_target = test_numerical_columns['SalePrice']

      # #Creating a linear regression instance  
        linear_model = LinearRegression()  
        linear_model.fit(train_features, train_target)
        predictions_one = linear_model.predict(test_features)        
        
        mse_one = mean_squared_error(test_target, predictions_one)
        rmse_one = np.sqrt(mse_one)
        
        linear_model.fit(train_features, train_target)
        predictions_two = linear_model.predict(test_features)        
       
        mse_two = mean_squared_error(test_target, predictions_two)
        rmse_two = np.sqrt(mse_two)
        
        avg_rmse = np.mean([rmse_one, rmse_two])
        print(rmse_one)
        print(rmse_two)
        return avg_rmse

  else:
        kf = KFold(n_splits=k, shuffle=True)
        rmse_values = []
        for train_index, test_index, in kf.split(data):
            train = data.iloc[train_index]
            test = data.iloc[test_index]
            linear_model.fit(train_features, train_target)
            predictions = linear_model.predict(test_features)
         

        mse = mean_squared_error(test_target, predictions)
        rmse = np.sqrt(mse)
        rmse_values.append(rmse)
        print(rmse_values)
        avg_rmse = np.mean(rmse_values)
        return avg_rmse


In [204]:
transform_data = transform_features(data)
filtered_data = select_features(transform_data)
rmse = train_and_test(filtered_data, k=3)
rmse

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:3069: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4174: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#retu

KeyError: ignored

## 5. Next Steps

That's it for the guided steps. Here's some potenial next steps that you can take:

1. Continue iteration on feature engineering:
* Research some other approaches to feature engineering online around housing data.
* Visit the Kaggle kernels [page](https://www.kaggle.com/c/house-prices-advanced-regression-techniques/kernels) page for this dataset to see approaches others took.

2. Improve your feature selection:
* Research ways of doing feature selection better with categorical columns (something we didn't cover in this particular course).